cpp/src/arrow/csv/column_decoder.cc

```diff
@@ -198,6 +198,11 @@ Result<std::shared_ptr<Array>> InferringColumnDecoder::RunInference(
 
 Future<std::shared_ptr<Array>> InferringColumnDecoder::Decode(
     const std::shared_ptr<BlockParser>& parser) {
+  if (parser->num_rows() == 0) {
```

Comment:
```suggestion
  // Note: Care should be taken to discard these batches as the data type will be null
  // and won't match future decoded arrays.
  if (parser->num_rows() == 0) {
```

Nit: Not necessary but might help explain to future readers.  Optionally you could also return `nullptr` here so it is clear it is invalid but then that will cause some confusion below between EOF and invalidly parsed batch.  I'll defer to whatever you think is best.